### PR TITLE
App Converter

### LINKED ISSUE #79: [Backend] Create App Converter

### PR DESCRIPTION
**Pull request by bhassan6621**
takes in an array of entities and converts them to apps

### CODE REVIEW

File: src/test/java/com/google/sps/service/AppReaderTest.java [hunk 1]
```diff
@@ -14,6 +14,7 @@
 
 package com.google.sps.service;
 
+import com.google.sps.models.App;
```
Comment (tonymai) [inline src/test/java/com/google/sps/service/AppReaderTest.java]: why is this needed?

File: src/test/java/com/google/sps/service/AppConverterTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Test
+    public void testAppReader() throws Exception {
+        AppSeeder populateDB = new AppSeeder();
+        populateDB.seedDB();
+
+        ArrayList<String> appsToQuery = new ArrayList<String>();
+        appsToQuery.add("com.google.android.wearable.app");
+        appsToQuery.add("com.acmeaom.android.myradar");
+        ArrayList<Entity> appEntities = new AppReader().getApps(appsToQuery);
```
Comment (tonymai) [inline src/test/java/com/google/sps/service/AppConverterTest.java]: Line 47-52 is doing a lot of set up. Is it possible to just create mocked `Entity` objects and pass it into the App Converter?

File: src/main/java/com/google/sps/service/AppConverter.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+           apps.add(convertEntityProperties(entity));
+        }
+        return apps;
+    }
+
+    public App convertEntityProperties(Entity entity) {
+        return addApp((String)entity.getProperty("id"), (String)entity.getProperty("title"), (String)entity.getProperty("url"), (String)entity.getProperty("category"), (double)entity.getProperty("rating"), (double)entity.getProperty("price"), (String)entity.getProperty("rated"), (String)entity.getProperty("description"));
+    }
+
+    public App addApp(String id, String title, String url, String category, double rating, double price, String rated, String appDescription) {
```
Comment (tonymai) [inline src/main/java/com/google/sps/service/AppConverter.java]: This method isn't doing much, I think it can be combined with the one above. Make it more clear with line breaks, i.e.

``` java
public App convertEntityProperties(Entity entity) {
  return new App.Builder()
    .setTitle(entity.getProperty("title"))
    .setUrl(entity.getProperty("url"))
    .setCategory(entity.getProperty("category"))
    ...
    .build();
}
```

**Review by tonymai — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/service/AppConverter.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.google.sps.service;
+
+import com.google.sps.models.App;
+import com.google.appengine.api.datastore.Entity;
+import java.util.ArrayList;
+class AppConverter {
```
Comment (ajwang0) [inline src/main/java/com/google/sps/service/AppConverter.java]: Add a new line inbetween the imports and the class

File: src/main/java/com/google/sps/service/AppConverter.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    
+    public ArrayList<App> convertToApp(ArrayList<Entity> appEntities) {
+        ArrayList<App> apps = new ArrayList<App>();
+        for(Entity entity: appEntities) {
+           apps.add(convertEntityProperties(entity));
+        }
+        return apps;
+    }
+
+    public App convertEntityProperties(Entity entity) {
```
Comment (ajwang0) [inline src/main/java/com/google/sps/service/AppConverter.java]: You can make this private

File: src/main/java/com/google/sps/service/AppConverter.java (hunk 3 above)
Comment (ajwang0) [inline src/main/java/com/google/sps/service/AppConverter.java]: You can make this private as well

File: src/main/java/com/google/sps/service/AppConverter.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package com.google.sps.service;
+
+import com.google.sps.models.App;
+import com.google.appengine.api.datastore.Entity;
+import java.util.ArrayList;
+class AppConverter {
+    
+    public ArrayList<App> convertToApp(ArrayList<Entity> appEntities) {
```
Comment (ajwang0) [inline src/main/java/com/google/sps/service/AppConverter.java]: Change ArrayList to just `List`

**Review by ajwang0 — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/models/App.java [hunk 7]
```diff
@@ -2,10 +2,41 @@
...
+
+        public String getCategory() {
+            return category;
+        }
+
+        public String getRated() {
+            return rated;
+        }
+
+        public String appDescription() {
```
Comment (tonymai) [inline src/main/java/com/google/sps/models/App.java]: prefix with `get`

File: src/main/java/com/google/sps/models/App.java [hunk 8]
```diff
@@ -2,10 +2,41 @@
...
+        }
+
+        public double getRating() {
+            return rating;
+        }
+
+        public double getPrice() {
+            return price;
+        }
+
```
Comment (tonymai) [inline src/main/java/com/google/sps/models/App.java:39]: please rebase, there are two more fields on this model now

File: src/test/java/com/google/sps/service/AppConverterTest.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        helper.setUp();
+    }
+
+    @AfterClass
+    public static void done() {
+        helper.tearDown();
+    }
+
+    @Test
+    public void testAppReader() throws Exception {
```
Comment (tonymai) [inline src/test/java/com/google/sps/service/AppConverterTest.java]: `testAppSeeder`

File: src/test/java/com/google/sps/service/AppConverterTest.java (hunk 2 above)
Comment (tonymai) [reply src/test/java/com/google/sps/service/AppConverterTest.java]: Doesn't look resolved

**Review by tonymai — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/sps/service/AppConverterTest.java (hunk 9 above)
Comment (ajwang0) [reply src/test/java/com/google/sps/service/AppConverterTest.java]: I think this one is testing the app reader since she's seeding the DB then using app reader to verify the data can be queried properly.

Comment (ajwang0) [reply src/test/java/com/google/sps/service/AppConverterTest.java]: Nvm ignore my comment i thought this was AppReaderTest

File: src/test/java/com/google/sps/service/AppConverterTest.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        app2.setProperty("title", "test app 2");
+        app2.setProperty("price", 0.99);
+        app2.setProperty("rating", 3.6); 
+        app2.setProperty("rated", "Everyone");
+        datastore.put(app2); 
+
+        entities.add(app1);
+        entities.add(app2);
+
+        List<App> convertedApps = new AppConverter().convertToApp(entities);
```
Comment (ajwang0) [inline src/test/java/com/google/sps/service/AppConverterTest.java:70]: Can you also verify there's 2 items in the list.

It might be better to construct your expected Apps and compare against that.

File: src/test/java/com/google/sps/service/AppReaderTest.java [hunk 11]
```diff
@@ -41,7 +41,7 @@ public static void done() {
     }
 
     @Test
-    public void testAppReader() throws Exception {
+    public void testAppConverter() throws Exception {
```
Comment (ajwang0) [inline src/test/java/com/google/sps/service/AppReaderTest.java]: This one should stay as testAppReader since this is testing the AppReader

**Review by ajwang0 — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/service/AppConverter.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package com.google.sps.service;
+
+import com.google.sps.models.App;
+import com.google.appengine.api.datastore.Entity;
+import java.util.ArrayList;
+import java.util.List;
+
+class AppConverter {
+    
+    public List<App> convertToApp(ArrayList<Entity> appEntities) {
```
Comment (ajwang0) [inline src/main/java/com/google/sps/service/AppConverter.java]: We should also make the input a List


`public List<App> convertToApp(List<Entity> appEntities) {`

**Review by ajwang0 — CHANGES_REQUESTED** [review]

**Review by ajwang0 — DISMISSED** [review]

File: src/test/java/com/google/sps/service/AppConverterTest.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+import org.junit.BeforeClass;
+import org.junit.Test;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public final class AppConverterTest {
+  private static final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig().setDefaultHighRepJobPolicyUnappliedJobPercentage(0));
```
Comment (ajwang0) [inline src/test/java/com/google/sps/service/AppConverterTest.java]: You can remove this too + the setup/done

**Review by ajwang0 — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/sps/service/AppConverterTest.java (hunk 13 above)
Comment (tonymai) [reply src/test/java/com/google/sps/service/AppConverterTest.java]: +1

**Review by tonymai — DISMISSED** [review]

**Review by ajwang0 — DISMISSED** [review]

File: src/test/java/com/google/sps/service/AppConverterTest.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public final class AppConverterTest {
+  // LocalServiceTestHelper is needed in order to avoid 'no API environment' error
```
Comment (tonymai) [inline src/test/java/com/google/sps/service/AppConverterTest.java:30]: `LocalServiceTestHelper` is needed for testing local App Engine services, but I don't think the `LocalDatastoreServiceTestConfig` needs to be passed in as an argument. Try without it.

**Review by tonymai — DISMISSED** [review]

**Review by ajwang0 — APPROVED** [review]

**Review by tonymai — APPROVED** [review]
